File: src/main/scala/com/formious/models/Participant.scala
package com.formious.models

import java.time.ZonedDateTime
import java.sql.ResultSet
import com.formious.common.Database.query
import com.formious.common.Recoders._

case class Participant(id: Int,
                       name: Option[String],
                       aws_worker_id: Option[String],
                       aws_bonus_owed: math.BigDecimal,
                       aws_bonus_paid: math.BigDecimal,
                       ip_address: Option[String],
                       user_agent: Option[String],
                       created: ZonedDateTime)

object Participant {
  def apply(row: ResultSet) = new Participant(
    row.getInt("id"),
    Option(row.getString("name")),
    Option(row.getString("aws_worker_id")),
    row.getBigDecimal("aws_bonus_owed"),
    row.getBigDecimal("aws_bonus_paid"),
    Option(row.getString("ip_address")),
    Option(row.getString("user_agent")),
    row.getTimestamp("created").toZonedDateTime)

  def find(id: Int) = {
    query("SELECT * FROM participant WHERE id = ?", List(id))(Participant(_)).head
  }

  def create(aws_worker_id: String, ip_address: Option[String], user_agent: Option[String]) = {
    query("""
      INSERT INTO participant (aws_worker_id, ip_address, user_agent)
      VALUES (?, ?, ?)
      RETURNING *
    """, List(aws_worker_id, ip_address, user_agent))(Participant(_)).head
  }

  def findOrCreateByWorkerId(aws_worker_id: String, ip_address: Option[String], user_agent: Option[String]) = {
    query("""
      SELECT * FROM participant WHERE aws_worker_id = ?
    """, List(aws_worker_id))(Participant(_)).headOption.getOrElse {
      create(aws_worker_id, ip_address, user_agent)
    }
  }

  def findOrCreate(participantIdOption: Option[Int], workerIdOption: Option[String]) = {
    (participantIdOption, workerIdOption) match {
      case (Some(participantId), _) => Some(find(participantId))
      case (_, Some(workerId)) => Some(findOrCreateByWorkerId(workerId, None, None))
      case _ => None
    }
  }
}

File: src/com/formious/db/participant.clj
(ns com.formious.db.participant
  (:require [com.formious.db :as db])
  (:import [java.time ZonedDateTime]))

; Int Option[String] Option[String] math.BigDecimal math.BigDecimal Option[String] Option[String] ZonedDateTime
(defrecord Participant [id name aws_worker_id aws_bonus_owed aws_bonus_paid ip_address user_agent created])

(defn row->Participant
  [row]
  (map->Participant (update row :created .toZonedDateTime)))

(defn all
  []
  (map row->AWSAccount (db/query "SELECT * FROM participant ORDER BY id ASC")))

(defn insert!
  ; (aws_worker_id: String, ip_address: Option[String], user_agent: Option[String])
  [row]
  (->> row (db/insert! "participant") row->AWSAccountAdministrator))

(defn find-by-id
  [id]
  (-> (db/query ["SELECT * FROM participant WHERE id = ?", id])
      first
      row->Participant))

(defn update!
  [id set-map]
  (db/update! "participant" set-map ["id = ?" id]))

(defn find-or-create-by-worker-id
  [aws_worker_id & {:keys [ip_address user_agent]}]
  (if-let [participants (db/query ["SELECT * FROM participant WHERE aws_worker_id = ?" aws_worker_id])]
    (-> participants first row->AWSAccount)
    (insert! {:aws_worker_id aws_worker_id :ip_address ip_address :user_agent user_agent})))

(defn find-or-create
  ; TODO: refactor this
  [id aws_worker_id]
  (if id
    (find-by-id id)
    (when aws_worker_id
      (find-or-create-by-worker-id aws_worker_id nil nil))))
